### ISSUE TITLE
Question: Tag ID hash resolution

### ISSUE THREAD

**Issue by phoerious** [lexbor/lexbor#176]
I have a few questions regarding the resolution of tag names. As far as I understand, there is a list of numeric tag IDs for all standard HTML5 tags here

https://github.com/lexbor/lexbor/blob/31270b1dfe8851f3779ca13d6efaae128178925f/source/lexbor/tag/const.h#L23-L222

from which a bunch of hash table entries is created here (both upper and lowercase):

https://github.com/lexbor/lexbor/blob/31270b1dfe8851f3779ca13d6efaae128178925f/source/lexbor/tag/res.h#L27-L425

Any additional tag names will be created dynamically and are stored in a separate hash table that is allocated together with the document. 

I can retrieve tag names by searching the hash table or by dereferencing the hash table entries directly via their tag IDs, which are either the array offsets in the default tag table or a direct pointer to the hash entry in the document's table. The function for retrieving tag names by ID is this:

https://github.com/lexbor/lexbor/blob/31270b1dfe8851f3779ca13d6efaae128178925f/source/lexbor/tag/tag.c#L62-L74

My questions concern mainly the lifetime of these objects:

1. I suppose the `tag_id >= LXB_TAG__LAST_ENTRY` check above is made under the (incorrect but probably safe) assumption that a pointer value will never be a low number. Is that correct?
2. Why does the function take a `lexbor_hash_t *hash` pointer, which is never used (since entries are referenced directly)? Legacy compatibility?
3. The default tag hash entries have a static lifetime and are never deallocated, which is indicated by the `readonly` struct flag. Correct?
4. I see that dynamic hash entries have a ref counter. Is this used anywhere and are dynamic hash entries being deallocated if the last tag using them is destroyed? From what I could see, this is planned, but not implemented right now.

The last question is particularly important, because it determines whether it is safe to keep using the `lxb_char_t*` pointer for a given tag name after any sort of DOM manipulation. The implications would be that

- Tag names with IDs less than `LXB_TAG__LAST_ENTRY` are always safe to use
- Tag names with IDs greater than `LXB_TAG__LAST_ENTRY` are safe to use only as long as the document exists and no tags were deleted from the DOM.

Are my assumptions correct? Anything important that I missed? Is the same mechanism used for other strings? I can see it being used for tag names, attribute names, and namespaces. Any other?

**Comment by lexborisov** [comment]
Hi @phoerious 

1. Yes
2. Yes, legacy.
3. Yes, always alive.
4. This is a groundwork for the future link counter for tags. Not implemented right now. Most likely this place will be changed. 

> The last question is particularly important, because it determines whether it is safe...

At the moment everything is correct.

But in the future:
1. A document can generate an infinite number of different tags, attributes.
2. All of them can be temporary.
3. It would not make sense to store them all in a hash table. Tags will be removed if they are not referenced, that is, if the tag is not present in the document.

As you already understood, in the future it will not be possible to just take `lxb_tag_data_t *` or an identifier. You will need to increment the link count if you don't want to run into garbage.

**Comment by phoerious** [comment]
OK, thanks. I'm writing a Rust API wrapper and this means I cannot simply borrow the char data, but have to copy them into a new owned String. Right now, I could potentially return a borrowed reference with the same lifetime as the tree itself, but if you plan to garbage-collect tag names in the future, this would break. I could perhaps bind the lifetime to the node itself instead of the full tree, but for that at least issue #132 needs to be fixed.

**Comment by phoerious** [comment]
@lexborisov I have a few more questions, this time regarding CSS selectors. The code and examples aren't entirely self-explanatory and there is no documentation at all. I have working code, but I think it's not quite optimal and I'm not entirely certain about the (required) lifetimes of objects.

This is what I do:

```rust
// ... DOM tree is created ...

let css_selectors = "foo, bar";
let cb = ...;          // Selection callback
let mut ctx = ...;     // Callback context
let root_node = ...;   // Root *mut lxb_dom_node_t for traversal

// (1) Create and init parse memory
let memory = lxb_css_memory_create();
lxb_css_memory_init(memory, 64);   // 64 is a hard-coded minimum somehow?

// (2) Create and init CSS parser
let parser = lxb_css_parser_create();
lxb_css_parser_init(parser, ptr::null_mut());
parser.selectors = ptr::null_mut();     // Ensure NULL

// ... DOM tree lives, user calls node.query_selector_all("...") ...

{
    // (3) (Re-)create CSS selectors
    if !(*parser).selectors.is_null() {
        lxb_css_selectors_destroy((*parser).selectors, true);
    }
    (*parser).selectors = lxb_css_selectors_create();
    lxb_css_selectors_init((*parser).selectors);
    
    // (4) Parse CSS expression
    let sel_list = lxb_css_selectors_parse(parser, css_selectors.as_ptr(), css_selectors.len());
    // (Error handling ...)
    
    // (5) Create LXB selectors
    let selectors = lxb_selectors_create();
    lxb_selectors_init(selectors);
    
    // (6) Find elements
    lxb_selectors_find(selectors, root_node, sel_list, cb, addr_of_mut!(ctx).cast());
    
    // (7) Destroy LXB selectors
    lxb_selectors_destroy(selectors, true);
    parser.selectors = ptr::null_mut();    // Ensure NULL
}

// ... DOM tree goes out of scope ...

// (8) Destroy CSS selectors if not already null
if !(*parser).selectors.is_null() {
    lxb_css_selectors_destroy((*parser).selectors, true);
    (*parser).selectors = ptr::null_mut();
}

// (9) Destroy CSS parser and memory
lxb_css_parser_destroy(parser, true);

// (10) Destroy CSS parser memory
lxb_css_memory_destroy(memory, true);
```

Parts of these statements are encapsulated in RAII structs, so I would like to know which resources can (or should) be reused and which need to be recreated for each CSS selection.

1. I initialize the CSS parser and memory (1), (2) only once for the lifetime of a DOM document and both are destroyed when the tree goes out of scope (8), (9), (10).
2. I recreate (5) and destroy (7) the LXB selectors for each selection process (4), (6).

Questions:

1. Do I need to recreate the LXB and / or CSS selectors for each new selection or can I reuse them (one or both) without leaking memory?
2. How exactly does `lxb_selectors_t` differ from `lxb_css_selectors_t`? Why do I need to create both explicitly?
3. Does the parser memory grow and shrink as needed? What happens if I exceed the initial number of `64` items? Do I leak memory by not destroying it every time?
4. Do I get any advantage from retaining the parser memory across multiple selections when the user's selectors are most likely different each time?
5. What is the lifetime of `sel_list`? In the "list_easy_way.c" example, you use `lxb_css_selector_list_destroy_memory()` to deallocate it explicitly, but in "list_fast_way.c" this is apparently handled by `lxb_css_memory_destroy()` (and I get a crash if I used both).
6. Do I even need `lxb_css_memory_destroy()` if I do `lxb_css_selectors_destroy()` after each selection, or vice versa: would it be better not to use `lxb_css_selectors_destroy()` and instead clean up in the end with `lxb_css_memory_destroy()` only?

Bonus question: Am I correct to assume that the parser returns a `lxb_css_selector_list_t`, because CSS selectors can be a comma-separated list of selectors and each entry represents an atomic selector?

**Comment by phoerious** [comment]
> You can create and initialize CSS Selectors before the loop. It would be very wasteful to do this every iteration of the loop.

So these handler objects are independent of the actual selectors. I checked the code again and `lxb_selectors_create()` and `lxb_css_selectors_create()` each do indeed only reserve space for a single object of this type. The handler itself is apparently constant and will only hold pointers to the actual individual selector objects. I don't see a reason to re-create them every time. Thanks!

> It depends on the architecture and how you will use it in the future.
If you have a lot of selectors, it's better to use one memory for all.

Many users will probably run only few selections on a document, but there can also be very heavy use of selectors in a loop.

> Undoubtedly. The parser doesn't care what the selectors are, and ultimately it has nothing to do with the result.

Understood.

> These are completely different entities.
> 
> The CSS contains modules: Syntax, Selectors, Values, Sizing, Box and so on. Each module is its own entity with its own rules.

I suspected something along the lines, but the details aren't really clear. I'm still not entirely sure I completely understand what each does. Since I always need both, it is just seems like very inconvenient and transparent boilerplate code which could probably be hidden away.

> 64 is pre-allocated memory for objects, text and so on. It is better to consider this place as pre-allocated memory for 64 `lxb_css_selector_t` objects. If the value is exceeded, an additional allocations of 64 objects will occur.

Thanks for confirming.

> When selectors are destroyed, their memory is cached and used for subsequent ones. In other words, a memory object will never return memory to the system, only after it is destroyed.

> After receiving the result of parsing (`lxb_css_selectors_parse()`), you can destroy all objects that are somehow related to parsing (except for memory). `sel_list` is completely independent.

From your explanation and from reading through the code, I understood that the memory is used for internal "dobjects" which I neither allocate nor deallocate directly myself. What is the primary scaling factor for these? Do I assume correctly that I will eventually bloat memory if I

- use very complex CSS selectors,
- do not free the internal objects with `lxb_css_memory_destroy()` after `lxb_selectors_find()`?

That would suggest that I should retain the parser memory and the CSS / LXB selector handler objects, but use `lxb_css_memory_destroy()` after each selection. Then memory usage would should be determined only by what my most complex selector requires but not by the number of selections in a loop, right?

**Comment by lexborisov** [comment]
@phoerious 

> use very complex CSS selectors,
> do not free the internal objects with lxb_css_memory_destroy() after lxb_selectors_find()?

Understood nothing.
Why are you parsing a lot of complicated selectors if you don't use or free them?

For example, if you parse CSS StyleSheet, then after you finish working with the result, simply destroy the memory object (`lxb_css_memory_destroy()`). Of course, you can go through all the selectors and destroy them manually (`lxb_css_selector_list_destroy()`), but it's easier to kill all the memory and create it again. It will be much faster and more efficient.

For example, if you have your own selectors for each iteration in the loop, and the previous ones are not needed, then simply destroy selectors (`lxb_css_selector_list_destroy()`). Memory will not grow, memory is reused. 

In general, there is a very flexible system for working with memory that allows you to adjust the use of memory to any situation.

**Comment by phoerious** [comment]
> Understood nothing. Why are you parsing a lot of complicated selectors if you don't use or free them?
> 
> For example, if you parse CSS StyleSheet, then after you finish working with the result, simply destroy the memory object (`lxb_css_memory_destroy()`). Of course, you can go through all the selectors and destroy them manually (`lxb_css_selector_list_destroy()`), but it's easier to kill all the memory and create it again. It will be much faster and more efficient.

Ah, no, we are talking about two different scenarios here. I am not parsing a larger style sheet (sorry for the confusion, in that case I would of course simply delete everything at once at the end). This is about a DOM API where a user can call a function such as `Element::querySelector()`. The calls should be stateless, so memory usage doesn't grow uncontrollably if the function is called in a loop. Everything belonging to that *particular* selector needs to be freed before returning the result.

> For example, if you have your own selectors for each iteration in the loop, and the previous ones are not needed, then simply destroy selectors (`lxb_css_selector_list_destroy()`). Memory will not grow, memory is reused.

That's pretty much the scenario. I would permanently (with the lifetime of the DOM tree object itself) allocate:

- CSS parser (`lxb_css_parser_create()`)
- CSS parser memory (`lxb_css_memory_create()`)
- CSS selectors handler (`lxb_css_selectors_create()`)
- LXB selectors handler (`lxb_selectors_create()`)

The selector list returned by `lxb_css_selectors_parse()` would be kept only temporarily (for the duration of the function call) and destroyed with `lxb_css_selector_list_destroy()` before returning the matching nodes.

Does that make sense?

**Comment by lexborisov** [comment]
@phoerious 

> Does that make sense?

Actually, [all CSS objects](https://github.com/lexbor/lexbor/blob/master/source/lexbor/html/interfaces/document.h#L44) for a document are created once. Destroyed when the document is destroyed.
But each CSS StyleSheet uses its own memory (lxb_css_memory_t).

**Comment by lexborisov** [comment]
@phoerious 

[There are examples](https://github.com/lexbor/lexbor/tree/master/examples/lexbor/styles).

> With those I would only need to call lxb_css_selector_list_destroy() after each selector. 

Yes.

**Comment by phoerious** [comment]
Great. Thanks!
